keycloak/keycloak
# External Infinispan as cache - Part 1

Part 1 includes

* New experimental feature to enable the new code
* New providers using RemoteCache only
* New test profile to run the tests with the experimental feature

New providers' implementation for:
* InfinispanConnectionProvider
* AuthenticationSessionProvider
* ClusterProvider

Closes #28140

---

## How to test

Spin up an external Infinispan, and redirect multicast discovery to a port different from the one used for the embedded Keycloak

```
podman run --replace -d --name infinispan \
--network host -e USER="keycloak" -e PASS="Password1!" \
quay.io/infinispan/server:14.0.27.Final \
-Djgroups.mcast_port=9999
```

Start up Keycloak to connect to the remote Infinispan. For development environments, add the additional Java options to disable TLS for a development environment, and to create the caches in the remote Infinispan if they don't exist yet. Both options are not supported for production environments, as TLS is a security feature, and remote caches should be configured with proper options like outlined in the Keycloak HA guide in an external infinispan.

```
export JAVA_OPTS_APPEND="-Dkc.cache-remote-tls-enabled=false -Dkc.cache-remote-create-caches=true"
bin/kc.sh start --http-enabled true --hostname-strict false  \
--features=multi-site,remote-cache \
--cache-remote-host localhost \
--spi-sticky-session-encoder-infinispan-should-attach-route=false \
--cache-remote-username=keycloak \
--cache-remote-password="Password1!"
```
